File: core/src/main/java/javabot/dao/ShunDao.java
package javabot.dao;

import java.time.LocalDateTime;

import javabot.model.Shun;
import javabot.criteria.ShunCriteria;

public class ShunDao extends BaseDao<Shun> {

  public ShunDao() {
    super(Shun.class);
  }

  public boolean isShunned(final String nick) {
    return getShun(nick) != null;
  }

  public Shun getShun(final String nick) {
    expireShuns();
    ShunCriteria criteria = new ShunCriteria(ds);
    criteria.upperNick().equal(nick.toUpperCase());
    return criteria.query().get();
  }

  private void expireShuns() {
    ShunCriteria criteria = new ShunCriteria(ds);
    criteria.expiry().lessThan(LocalDateTime.now());
    ds.delete(criteria.query());
  }

  public void addShun(final String nick, final LocalDateTime until) {
    Shun shun = getShun(nick);
    if (shun == null) {
      shun = new Shun();
      shun.setNick(nick.toUpperCase());
      shun.setExpiry(until);
      save(shun);
    }
  }
}

File: src/main/kotlin/javabot/dao/ShunDao.kt
package javabot.dao

import com.google.inject.Inject
import dev.morphia.Datastore
import dev.morphia.DeleteOptions
import dev.morphia.query.filters.Filters.eq
import dev.morphia.query.filters.Filters.lt
import java.time.LocalDateTime
import java.util.Locale
import javabot.model.Shun

class ShunDao @Inject constructor(ds: Datastore) : BaseDao<Shun>(ds, Shun::class.java) {

    fun isShunned(nick: String): Boolean {
        return getShun(nick) != null
    }

    fun getShun(nick: String): Shun? {
        expireShuns()
        return ds.find(Shun::class.java)
            .filter(eq("upperNick", nick.uppercase(Locale.getDefault())))
            .first()
    }

    private fun expireShuns() {
        ds.find(Shun::class.java)
            .filter(lt("expiry", LocalDateTime.now()))
            .delete(DeleteOptions().multi(true))
    }

    fun addShun(nick: String, until: LocalDateTime) {
        var shun = getShun(nick)
        if (shun == null) {
            shun = Shun()
            shun.nick = nick.uppercase(Locale.getDefault())
            shun.expiry = until
            save(shun)
        }
    }
}
